parl/framework/algorithm.py

```diff
@@ -0,0 +1,188 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle.fluid as fluid
+import parl.layers as layers
+from parl.layers import Network
+import parl.framework.policy_distribution as pd
+
+
+def check_duplicate_spec_names(model):
+    """
+    Check if there are two specs that have the same name.
+    """
+    specs = model.get_input_specs() \
+            + model.get_action_specs() \
+            + model.get_state_specs() \
+            + model.get_reward_specs()
+    names = [name for name, _ in specs]
+    duplicates = set([n for n in names if names.count(n) > 1])
+    assert not duplicates, \
+        "duplicate names with different specs: " + " ".join(duplicates)
+
+
+class Model(Network):
+    """
+    A Model is owned by an Algorithm. It implements all the network model of
+    a specific problem.
+    """
+
+    def __init__(self):
+        super(Model, self).__init__()
+
+    def get_input_specs(self):
+        """
+        Output: list of tuples
+        """
+        raise NotImplementedError()
+
+    def get_state_specs(self):
+        """
+        States are optional to a Model.
+        Output: list of tuples
+        """
+        return []
+
+    def get_action_specs(self):
+        """
+        Output: list of tuples
+        """
+        raise NotImplementedError()
+
+    def get_reward_specs(self):
+        """
+        By default, a scalar reward.
+        User can specify a vector of rewards for some problems
+        """
+        return [("reward", dict(shape=[1]))]
+
+    def policy(self, inputs, states):
```

Comment:
Is this function used to do prediction in supervised learning task?

Reply:
Yes.

Reply:
that means ct has to be changed when using supervised learning?

Reply:
ct doesn't have to be changed. it always call alg.predict and inside that the user decides which function of the model to call